zebra-network/src/peer_set/candidate_set.rs

```diff
@@ -329,24 +332,52 @@ where
 /// If the data in an address is invalid, this function can:
 /// - modify the address data, or
 /// - delete the address.
-//
-// TODO: re-enable this lint when last_seen_limit is used
-#[allow(unused_variables)]
+///
+/// # Security
+///
+/// Adjusts untrusted last seen times so they are not in the future. This stops
+/// malicious peers keeping all their addresses at the front of the connection
+/// queue. Honest peers with future clock skew also get adjusted.
 fn validate_addrs(
     addrs: impl IntoIterator<Item = MetaAddr>,
     last_seen_limit: DateTime<Utc>,
-) -> impl IntoIterator<Item = MetaAddr> {
+) -> impl Iterator<Item = MetaAddr> {
     // Note: The address book handles duplicate addresses internally,
     // so we don't need to de-duplicate addresses here.
 
     // TODO:
     // We should eventually implement these checks in this function:
-    // - Zebra should stop believing far-future last_seen times from peers (#1871)
     // - Zebra should ignore peers that are older than 3 weeks (part of #1865)
     //   - Zebra should count back 3 weeks from the newest peer timestamp sent
     //     by the other peer, to compensate for clock skew
     // - Zebra should limit the number of addresses it uses from a single Addrs
     //   response (#1869)
 
-    addrs
+    let mut addrs: Vec<_> = addrs.into_iter().collect();
+
+    if !addrs.is_empty() {
+        limit_last_seen_times(&mut addrs, last_seen_limit);
+    }
+
+    addrs.into_iter()
+}
+
+/// Ensure all reported `last_seen` times are less than or equal to `last_seen_limit`.
+///
+/// # Panics
+///
+/// If the `addrs` list is empty.
+fn limit_last_seen_times(addrs: &mut Vec<MetaAddr>, last_seen_limit: DateTime<Utc>) {
+    let most_recent_reported_seen_time = addrs
+        .iter()
+        .map(|addr| addr.get_last_seen())
+        .max()
+        .expect("unexpected empty address list");
+
+    if most_recent_reported_seen_time > last_seen_limit {
```

Comment:
Let's explain why we chose this alternative:
```suggestion
    // If any time is in the future, adjust all times, to compensate for clock skew on honest peers
    if most_recent_reported_seen_time > last_seen_limit {
```